pkg/upgrade/upgrade.go

```diff
@@ -0,0 +1,357 @@
+			return fmt.Errorf("running upgrade binary %s: %w", binary, err)
+		}
+
+		// Mark as run
+		err = u.markAsRun(binary)
+		if err != nil {
+			return fmt.Errorf("marking upgrades as run: %w", err)
+		}
+
+		// Update cluster version
+		// Note that we don't save the hotfix version, only the semver version. This is because we use the cluster
+		// versioner later to validate that the current okctl version against the cluster version. The okctl version
+		// will always be a semver version without any hotfix in it, so it wouldn't be possible to store a hotfix
+		// version here.
+		//
+		// Also, a hotfix version is supposed to just fix any errors being made in an upgrade binary, so
+		// the effect of an upgrade binary version 0.0.10 plus a hotfix binary 0.0.10.a, should be as if running one
+		// working upgrade binary with version 0.0.10.
+		err = u.clusterVersioner.SaveClusterVersion(binary.SemverVersion().String())
+		if err != nil {
+			return fmt.Errorf(commands.SaveClusterVersionError, err)
+		}
+	}
+
+	return nil
+}
+
+func (u Upgrader) createBinaryProvider(upgradeBinaries []okctlUpgradeBinary) (upgradeBinaryProvider, error) {
+	binaries := u.toStateBinaries(upgradeBinaries)
+
+	// fetch.New downloads binaries that are missing
+	fetcher, err := fetch.New(
+		u.out,
+		u.logger,
+		true,
+		u.fetcherOpts.Host,
+		binaries,
+		u.fetcherOpts.Store,
+	)
+	if err != nil {
+		return upgradeBinaryProvider{}, fmt.Errorf("creating upgrade binaries fetcher: %w", err)
+	}
+
+	binaryProvider := newUpgradeBinaryProvider(u.repositoryDirectory, u.logger, u.out, fetcher)
+
+	return binaryProvider, nil
+}
+
+func (u Upgrader) dryRunBinaries(upgradeBinaries []okctlUpgradeBinary, binaryProvider upgradeBinaryProvider) error {
+	_, _ = fmt.Fprint(u.out, "Simulating upgrades...\n\n")
+
+	for _, binary := range upgradeBinaries {
+		// Get
+		binaryRunner, err := binaryProvider.okctlUpgradeRunner(binary.RawVersion())
+		if err != nil {
+			return fmt.Errorf("getting okctl upgrade binary: %w", err)
+		}
+
+		binaryRunner.SetDebug(u.debug)
+
+		_, _ = fmt.Fprintf(u.out, "--- Simulating upgrade: %s ---\n", binary)
+
+		// Run
+		_, err = binaryRunner.Run(false)
+		if err != nil {
+			_, _ = fmt.Fprintf(u.out, "--- Upgrade failed: %s ---\n", binary)
+			return fmt.Errorf("running upgrade binary %s: %w", binary, err)
+		}
+	}
+
+	_, _ = fmt.Fprintf(u.out, "\nSimulating upgrades complete.\n\n")
+
+	return nil
+}
+
+func (u Upgrader) toStateBinaries(upgradeBinaries []okctlUpgradeBinary) []state.Binary {
+	binaries := make([]state.Binary, 0, len(upgradeBinaries))
+
+	for _, upgradeBinary := range upgradeBinaries {
+		URLPattern := fmt.Sprintf(
+			"https://github.com/oslokommune/okctl-upgrade/releases/download/%s/okctl-upgrade_%s_#{os}_#{arch}.tar.gz",
+			upgradeBinary.RawVersion(),
+			upgradeBinary.RawVersion(),
+		)
+
+		binary := state.Binary{
+			Name:       upgradeBinary.BinaryName(),
+			Version:    upgradeBinary.RawVersion(),
+			BufferSize: "300mb",
```

Comment:
Where is this number from? Can/will this bite us in the ass?

Reply:
It could be 10mb and still work (but more overhead), a buffer is just the size of temporary storage. Given that it was implemented correctly.

I'm following the same system as ekctl, kubectl and aws-iam-authenticator, but they use `100mb`. I assumed upgrades could be larger, so I upped it to 300mb.